modules/recon/credentials-credentials/bozocrack.py

```diff
@@ -38,10 +39,16 @@ def module_run(self, hashes):
             # sleep to avoid lock-out
             time.sleep(random.randint(3,5))
 
+
 def crack(hashstr, wordlist):
     for word in wordlist:
-        for hashtype in hashlib.algorithms:
-            func = getattr(hashlib, hashtype)
-            if func(word).hexdigest().lower() == hashstr.lower():
-                return word, hashtype
+        if word != '':
+            for hashtype in hashlib.algorithms_guaranteed:
+                h = hashlib.new(hashtype)
+                h.update(word.encode('utf-8'))
+                try:
+                    if h.hexdigest().lower() == hashstr.lower():
+                        return word, hashtype
+                except TypeError:
```

Comment:
What was causing this error? I'm not sure how you would get a `TypeError` here from something that is completely text. If it is an encoding issue, then I have a built-in method that will detect and convert to the proper encoding.

Reply:
This error was being caused because on certain devices (I'm testing from a Macbook Pro) when you run hashlib.algorithms_guaranteed you get back two algos shake_128() and shake_256(). 

The shake_128() and shake_256() algorithms provide variable length digests with length_in_bits//2 up to 128 or 256 bits of security. As such, their digest methods require a length

So you have to use it h.hexdigest(length) unlike the other algorithms.

Reply:
I was also looking at the supported hashes you have listed in the description. Seems that hashlib doesn't always have those available actually none of my boxes have those. Wondering if we should implement this a bit differently. Building a function that handles the other functions such as NTLM which are not available anymore.

Reply:
Saw a comment in email that didn't show up here for some reason...

The listed hashing algorithms are no longer available? That seems like an odd change to make to a hashing library, to remove all of the algorithms.

Reply:
This is what mine shows. Perhaps we go with `available` then make a note that says something like "Supported algorithms depend on those available on the current system." Or, make the not dynamic and do something like `f"Supported algorithms: {', '.join(hashlib.algorithms_available)}"`

Thoughts?

```
>>> hashlib.algorithms_guaranteed
{'blake2s', 'md5', 'sha512', 'blake2b', 'sha3_256', 'sha3_512', 'sha3_384', 'shake_128', 'sha256', 'sha224', 'sha1', 'sha3_224', 'shake_256', 'sha384'}
>>> hashlib.algorithms_available
{'SHA1', 'MD5', 'sha3_384', 'shake_128', 'SHA', 'blake2s', 'RIPEMD160', 'sha512', 'ripemd160', 'sha3_256', 'sha3_512', 'mdc2', 'sha3_224', 'shake_256', 'MD4', 'SHA512', 'SHA256', 'MDC2', 'blake2b', 'ecdsa-with-SHA1', 'DSA-SHA', 'sha256', 'SHA224', 'sha224', 'sha384', 'DSA', 'md5', 'sha', 'dsaEncryption', 'whirlpool', 'dsaWithSHA', 'sha1', 'md4', 'SHA384'}
```

Reply:
I typed all that out realizing I was looking at a different module as I made all those comments so I removed the comment. I think available could lead to us having to make a lot more error checking (Would need to test this just assuming at the moment). With guaranteed I think we are avoiding having to add extra error checking than is really needed.

Reply:
I've edited my comment a couple times. Thoughts on the 3 proposed options?

Reply:
I like putting the comment about supported algorithms but I think we should stick with hashlib.algorithms_guaranteed I think it will make the module more stable.

Reply:
I agree with that as well, but what do you think about having a default subset of `hashlib.algorithms_available` set as an option with the ability to modify? My reasoning is that quite often when reversing hashes, we either know the algorithm used, or know which algorithms are likely. So rather than force all algorithms for every guess, we could save a bunch of time with a configurable, preset list of algorithms. We would just need to bounce the provided list off of those available before processing to make sure someone doesn't try to use an unsupported algorithm. Thoughts?

Reply:
Ahhhh ya gotcha now. Yep, I like that. I have some airport time later today I can toss that together pretty easily.

Reply:
So just updated this one... I'm thinking this could be a bit cleaner so will take any recommendations but its functional and works.

Reply:
Make some tweaks. Please test it and let me know if it works as is. If so, it's good to go.